lib/ansible/modules/mount_facts.py

```diff
@@ -0,0 +1,651 @@
+        mount_info: dict[str, str | int] = {
+            "device": fields[0],
+            "mount": fields[1],
+            "fstype": fields[2],
+            "options": fields[3],
+            "time": int(fields[4]),
+        }
+        yield MountInfo(fields[1], line, mount_info)
+
+
+def gen_mounts_by_file(file: str) -> t.Iterable[MountInfo | MountInfoOptions]:
+    """Yield parsed mount entries from the first successful generator.
+
+    Generators are tried in the following order to minimize false positives:
+    - /etc/vfstab: 7 columns
+    - /etc/mnttab: 5 columns (mnttab[4] must contain UNIX timestamp)
+    - /etc/fstab: 4-6 columns (fstab[4] is optional and historically 0-9, but can be any int)
+    - /etc/filesystems: multi-line, not column-based, and specific to AIX
+    """
+    if (lines := get_file_content(file, "").splitlines()):
+        for gen_mounts in [gen_vfstab_entries, gen_mnttab_entries, gen_fstab_entries, gen_aix_filesystems_entries]:
+            with suppress(IndexError, ValueError):
+                # mpypy error: misc: Incompatible types in "yield from" (actual type "object", expected type "Union[MountInfo, MountInfoOptions]
+                # only works if either
+                # * the list of functions excludes gen_aix_filesystems_entries
+                # * the list of functions only contains gen_aix_filesystems_entries
+                yield from list(gen_mounts(lines))  # type: ignore[misc]
+                break
+
+
+def get_sources(module: AnsibleModule) -> list[str]:
+    """Return a list of filenames from the requested sources."""
+    sources: list[str] = []
+    for source in module.params["sources"] or ["all"]:
+        if not source:
+            module.fail_json(msg="sources contains an empty string")
+
+        if source in {"dynamic", "all"}:
+            sources.extend(DYNAMIC_SOURCES)
+        if source in {"static", "all"}:
+            sources.extend(STATIC_SOURCES)
+
+        elif source not in {"static", "dynamic", "all"}:
+            sources.append(source)
+    return sources
+
+
+def gen_mounts_by_source(module: AnsibleModule):
+    """Iterate over the sources and yield tuples containing the source, mount point, source line(s), and the parsed result."""
+    sources = get_sources(module)
+
+    if len(set(sources)) < len(sources):
+        module.warn(f"mount_facts option 'sources' contains duplicate entries, repeat sources will be ignored: {sources}")
+
+    mount_fallback = module.params["mount_binary"] and set(sources).intersection(DYNAMIC_SOURCES)
+
+    seen = set()
+    for source in sources:
+        if source in seen:
+            continue
+        seen.add(source)
+
+        if source == "mount":
+            stdout = run_mount_bin(module, module.params["mount_binary"])
+            results = [(source, *astuple(mount_info)) for mount_info in gen_mounts_from_stdout(stdout)]
+        else:
+            seen.add(os.path.realpath(source))
```

Comment:
If different symlinks refer to the same source, then the `source in seen` check above won't detect the duplicate. Checking the real source here would protect against that:

```suggestion
            if (real_source := os.path.realpath(source)) in seen:
                continue
                
            seen.add(real_source)
```

Reply:
I just realized the change I suggested doesn't resolve the issue I pointed out, since it would result in skipping a path that isn't a symlink. The seen check needs to be done in one place, or a separate set used for real paths.

Reply:
I have a warning above if non-symlink duplicates are given, so I think skipping those too is correct. How does https://github.com/ansible/ansible/pull/83508/commits/ef1f410f02f0a78de307cb9f403791b9241ebb55 look?